### ISSUE TITLE
Type inference of closure types isn't always handled properly

### ISSUE THREAD

**Issue by yorickpeterse** [inko-lang/inko#611]
### Please describe the bug

When returning a generic type in a closure without an explicit return type, the return type isn't always inferred correctly.

I'm not sure what exactly is happening just yet, but I'm guessing we end up comparing two unassigned type placeholders, and thus don't assign either of them, but that's just a guess at this point.

### Please list the exact steps necessary to reproduce the bug

Run this program:

```inko
class Thing {}

class async Main {
  fn async main {
    let a = [Thing {}]
    let b = a.into_iter.select_map fn (t) { Option.None }.to_array

    b.TEST
  }
}
```

This _should_ emit an error along the lines of this:

```
error(invalid-symbol): the method 'TEST' isn't defined for type 'Array[Thing]'
```

Instead, the error is this:

```
error(invalid-symbol): the method 'TEST' isn't defined for type 'Array[T]'
```

### Operating system

Fedora Silverblue 38

### Inko version

main

### Rust version

1.72.1

**Comment by yorickpeterse** [comment]
Looking at this again, the expected outcome isn't correct: the `select_map` method returns `Option.None` without ever specifying what the `T` in `Option[T]` is. The result is that the closure passed to `select_map` is inferred as `fn (Thing) -> Option[<unknown>]` where `<unknown>` indicates an unknown type. If we remove the `TEST` call the compiler indeed produces the expected inference errors.

In other words, I was mistaken when creating this issue, perhaps due to the other inference issues that I ran into (and fixed recently).
